Anti-virus support without need to run external programs

azurit opened the pull request:
Adding anti-virus support implemented in pure Lua. The only external requirement is lua-socket library (should be part of all major distros). Currently, only ClamAV is supported. All configuration can be done in crs-setup.conf. This feature is disabled by default.

Feature is implemented using `INSTREAM` ClamAV command, so:
 - no external program/script is running (no forking, no need of extended permissions and so on)
 - ClamAV doesn't need to be run as root to have access to scanned files as data is send through socket (both TCP/IP and unix socket files are supported)
 - should be fully multiplatform

Thing to consider: Requests containing virus should be hard-blocked, current implementation is only adding score (CRITICAL severity).

ClamAV signatures for PHP malware (tested, both are good):
https://www.rfxn.com/projects/linux-malware-detect/ (free of charge)
https://malware.expert/signatures/ (commercial)

dune73 commented:
Hey @azurit, thank you for this contribution.

I can not get it to work so far, but I have some preliminary feedback:

* The log action missing in rule 925100
* I like your handling of virus name in 925100
* Please rename av.lua to anti-virus.lua
* I get the feeling you would like to generalize this, but the var names link it to clamav. Can you generalize them, or will a different backend need different variables? If yes, then better leave it as is.
* I enabled the rules without compiling lua into ModSec, but no warning. Probably not your fault, but it's a problem.-
                (As soon as lua is installed, apache complains when it does not find the script.)
* No comments whatsoever in av.lua script. Most is self-explanatory, but not everything.
* Documentation should reference the need for testing with the eicar file:
                curl http://localhost --form "data=@eicar.com"
                Get eicar test file from https://www.eicar.org
 * Remark to run lua script from the command line would also be helpful
                lua5.3 av.lua 
                (If that is meant to work at all)
* No complaint that lua-socket is not installed
                But error message:
                        [2021-01-25 15:31:53.679752] [-:error] 127.0.0.1:44956 YA7WWVYoCzdOV1uTMfDC7AAAAAA [client 127.0.0.1] ModSecurity: Rule processing failed (id=, msg=). [hostname "localhost"] [uri "/"] [unique_id "YA7WWVYoCzdOV1uTMfDC7AAAAAA"]
* configuration comments reference connection type socket and tcp, but script checks for socket and else.
                Also not sure that 
                        module_name = "socket" 
                is correct for connection type "tcp". Or is it?
* What is require("m") meant to load? I fail with this. And I never got it to log anything.

azurit commented:
> Hey @azurit, thank you for this contribution.
> 
> I can not get it to work so far, but I have some preliminary feedback:


Thanks for reviewing! Strange, it's fully working for me. Also see comments below.


>     * The log action missing in rule 925100


Fixed.


>     * I like your handling of virus name in 925100
> 
>     * Please rename av.lua to anti-virus.lua


Done.


>     * I get the feeling you would like to generalize this, but the var names link it to clamav. Can you generalize them, or will a different backend need different variables? If yes, then better leave it as is.


I was planning to add support also for other anti-virus software and allow activation of multiple anti-viruses simultaneously - it makes sense as various AVs are able to detect different viruses.


>     * I enabled the rules without compiling lua into ModSec, but no warning. I know, this is not your fault, but it's a problem.
>       (As soon as lua is installed, apache complains when it does not find the script.)


Now i noticed that `@inspectFile` is not giving a warning about this - in other cases like `SecRuleScript`, modsecurity is printing a warning on web server startup:
`Ignoring SecRuleScript "response_body_decompress.lua" directive (/usr/share/modsecurity-crs-new/rules/REQUEST-901-INITIALIZATION.conf:502): No Lua scripting support.`

Probably a modsecurity bug, I will look into this.


>     * No comments whatsoever in av.lua script. Most is self-explanatory, but not everything.


I already added few but was not pushing it only because of this. But ok, i will add more.


>     * Documentation should reference the need for testing with the eicar file:
>       curl http://localhost --form "data=@eicar.com"
>       Get eicar test file from https://www.eicar.org


Added.


>     * Remark to run lua script from the command line while testing would also be helpful
>       lua5.3 av.lua
>       (If that is meant to work at all)


This is not supposed to work.


>     * No complaint that lua-socket is not installed
>       But error message:
>       [2021-01-25 15:31:53.679752] [-:error] 127.0.0.1:44956 YA7WWVYoCzdOV1uTMfDC7AAAAAA [client 127.0.0.1] ModSecurity: Rule processing failed (id=, msg=). [hostname "localhost"] [uri "/"] [unique_id "YA7WWVYoCzdOV1uTMfDC7AAAAAA"]


It's working for me, modsec2, apache 2.4, lua 5.1. What version of modsec are you testing on? Which web server software?

```
[Mon Jan 25 17:06:01.153312 2021] [:error] [pid 14721] [client ...] [client ...] ModSecurity: ClamAV: lua-socket library not installed, please install it or disable 'tx.crs_anti-virus_protection_enable' in crs-setup.conf [hostname "..."] [uri "/"] [unique_id "YA7saWbMVoYLfTj0kRzgpAAAAGA"]

```

>     * configuration comments reference connection type socket and tcp, but script checks for socket and else.


Yes, do you find this not correct? I can add additional check.


>       Also not sure that
>       module_name = "socket"
>       is correct for connection type "tcp". Or is it?


This is ok.


>     * What is require("m") meant to load? I fail with this. And I never got it to log anything.


It's a special modsecurity Lua library shipped with modsecurity and always available to all Lua scripts executed internally, using it, Lua scripts has access to modsecurity internals (for example `TX` variable).

dune73 commented:
Thank you for petting me. I have lua 5.3 installed. Does it work for you on lua 5.3 as well?

Definitely no "m" to load / require here. :(

azurit commented:
Are you sure your modsecurity is compiled with Lua 5.3? Is it modsec 2 or 3? The whole modsecurity 2.x documentation is mentioning only Lua version 5.1. Newer Lua versions are available only with modsec 3 (as far as i understand it).

dune73 commented:
Ah, my bad. Have not worked with Lua in quite a long time. No problem during compilation, but if that is a constraint, I need to downgrade and try it out anew.

azurit commented:
I just pushed mentioned fixies.

azurit commented:
Were you able to make it work?

dune73 commented:
Have not tried it yet.

franbuehler commented:
Meeting decision March: https://github.com/coreruleset/coreruleset/issues/2008#issuecomment-788239186:
@airween volunteers to review this PR.

dune73 commented:
Hey @airween: You promised to review this PR. Did you have the time to look into it?

airween commented:
> Hey @airween: You promised to review this PR. Did you have the time to look into it?

Yes, I already did it - found some issue in case of libmodsecurity3, see [this](https://github.com/SpiderLabs/ModSecurity-nginx/issues/244) report. Noticed @azurit about partially results.

Btw: in case of Apache2 - mod_security everything worked as well.

airween commented:
I reviewed this PR, see my comments between lines.

There is one more note: libmodsecurity3 has a [known](https://github.com/SpiderLabs/ModSecurity/pull/2045) [bug](https://github.com/SpiderLabs/ModSecurity/issues/2146): variable `REQUEST_BODY` always presents, even if the `Content-Type` **is not** `www-url-form-encoded` nor `ctl:forceRequestBodyVariable` **isn't used**. Therefore it has an unexpected side effect: when upload a file, libmodsecurity3 uses the `REQUEST_BODY` at all rules where it uses, and there are two false positive rule: [920272](https://github.com/coreruleset/coreruleset/blob/v3.4/dev/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf#L1362-L1378) and [920273](https://github.com/coreruleset/coreruleset/blob/v3.4/dev/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf#L1540-L1556).

```
2021/04/04 21:23:19 [info] 12723#12723: *3 ModSecurity: Warning. Matched "Operator `ValidateByteRange' with parameter `32-36,38-126' against variable `REQUEST_BODY' (Value: `--------------------------cbd70b742c11bc6b\x0d\x0aContent-Disposition: form-data; name="userfile"; f (556 characters omitted)' ) [file "/usr/share/modsecurity-crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "1361"] [id "920272"] [rev ""] [msg "Invalid character in request (outside of printable chars below ascii 127)"] [data "REQUEST_BODY=--------------------------cbd70b742c11bc6b\x0d\x0aContent-Disposition: form-data; name="userfile"; filename="eicar.com"\x0d\x0aContent-Type: application/octet-stream\x0d\x0a\x0d\x0aPK\x03\x04\x0a\x00\x00\x00\x00\x00\xe0\x98\xb8(<\xcfQhD\x00\x00\x00D\x00\x00\x00\x09\x00\x00\x00 (202 characters omitted)"] [severity "2"] [ver "OWASP_CRS/3.3.0"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "OWASP_CRS"] [tag "capec/1000/210/272"] [tag "paranoia-level/3"] [hostname "::1"] [uri "/index.php"] [unique_id "1617564199"] [ref "o42,1o43,1o113,1o114,1o153,1o154,1o155,1o156,1o159,1o160,1o161,1o162,1o163,1o164,1o165,1o166,1o167,1o168,1o169,1o172,1o176,1o177,1o178,1o180,1o181,1o182,1o183,1o184,1o185,1o186,1o201,1o266,1o267,1o268 (360 characters omitted)"], client: ::1, server: _, request: "POST /index.php HTTP/1.1", host: "localhost"

2021/04/04 21:23:19 [info] 12723#12723: *3 ModSecurity: Warning. Matched "Operator `ValidateByteRange' with parameter `38,44-46,48-58,61,65-90,95,97-122' against variable `REQUEST_BODY' (Value: `--------------------------cbd70b742c11bc6b\x0d\x0aContent-Disposition: form-data; name="userfile"; f (556 characters omitted)' ) [file "/usr/share/modsecurity-crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf"] [line "1500"] [id "920273"] [rev ""] [msg "Invalid character in request (outside of very strict set)"] [data "REQUEST_BODY=--------------------------cbd70b742c11bc6b\x0d\x0aContent-Disposition: form-data; name="userfile"; filename="eicar.com"\x0d\x0aContent-Type: application/octet-stream\x0d\x0a\x0d\x0aPK\x03\x04\x0a\x00\x00\x00\x00\x00\xe0\x98\xb8(<\xcfQhD\x00\x00\x00D\x00\x00\x00\x09\x00\x00\x00 (202 characters omitted)"] [severity "2"] [ver "OWASP_CRS/3.3.0"] [maturity "0"] [accuracy "0"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-protocol"] [tag "OWASP_CRS"] [tag "capec/1000/210/272"] [tag "paranoia-level/4"] [hostname "::1"] [uri "/index.php"] [unique_id "1617564199"] [ref "o42,1o43,1o64,1o74,1o75,1o81,1o90,1o91,1o92,1o102,1o112,1o113,1o114,1o128,1o140,1o153,1o154,1o155,1o156,1o159,1o160,1o161,1o162,1o163,1o164,1o165,1o166,1o167,1o168,1o169,1o170,1o171,1o172,1o176,1o177, (533 characters omitted)"], client: ::1, server: _, request: "POST /index.php HTTP/1.1", host: "localhost"
```

These FP's always triggered, no matter the anti-virus feature is on or off, but the anomaly score value will be distorted.

And finally: in case of Apache2 + mod_security everything is fine.

airween reviewed (commented):
See my comments between the lines.

azurit commented on crs-setup.conf.example:
```diff
@@ -752,6 +752,52 @@ SecDefaultAction "phase:2,log,auditlog,pass"
+# Currently, only ClamAV anti-virus is supported.
+#
+# Prerequisities for this feature:
+# - Lua support must be enabled
```
What is `ngx_mod_http_lua`? Some kind of Lua library?

airween replied:
Yes, it's an embedded Lua module:

https://github.com/openresty/lua-nginx-module

Name of module can be different by distributor - here is how Debian uses:
```
root@debian10:/etc/nginx/modules-enabled# cat 50-mod-http-lua.conf 
load_module modules/ngx_http_lua_module.so;
```

azurit replied:
Is this it?
https://github.com/openresty/lua-nginx-module

azurit replied:
Oh, sorry, didn't noticed you already posted the URL.

azurit commented on crs-setup.conf.example:
```diff
@@ -752,6 +752,52 @@ SecDefaultAction "phase:2,log,auditlog,pass"
+# - Lua support must be enabled
+# - Library lua-socket must be installed in the system
+# - modsecurity 'SecTmpSaveUploadedFiles' directive is 'On' or
+#   the 'SecUploadKeepFiles' directive is set to 'RelevantOnly'
```
`SecUploadKeepFiles` must be set to `RelevantOnly` or `On`, i will update the comment.

airween commented on crs-setup.conf.example:
```diff
@@ -752,6 +752,52 @@ SecDefaultAction "phase:2,log,auditlog,pass"
+#  setvar:'tx.crs_anti-virus_clamav_port=3310',\
+#  setvar:'tx.crs_anti-virus_clamav_chunk_size=4096',\
+#  setvar:'tx.crs_anti-virus_clamav_network_timeout_seconds=2',\
+#  setvar:'tx.crs_anti-virus_clamav_max_file_size_bytes=1048576'"
```
To avoid the FP's in case of libmodsecurity3 (because the `REQUEST_BODY` is always presented), we should add these actions:
```
  ctl:ruleRemoveById=920272,\
  ctl:ruleRemoveById=920273"
```
or we can combine these exceptions with checking of CT header.

azurit replied:
I would not count on CT as it can be faked to bypass anti-virus check.

airween replied:
May be we should discuss this on monthly chat. I tried this rule in crs-setup.conf:
```
SecRule REQUEST_HEADERS:Content-Type "(?:multipart/form-data)" \
    "id:'900801',\
    phase:1,\
    t:none,t:lowercase,\
    pass,\
    nolog,\
    ctl:ruleRemoveTargetById=920272;REQUEST_BODY,\
    ctl:ruleRemoveTargetById=920273;REQUEST_BODY"
```

azurit replied:
I will keep this open, i think we need to discuss it with others.

azurit commented on rules/anti-virus.lua:
```diff
@@ -0,0 +1,106 @@
+require("m")
```
Can you, please, test this now in modsec3?

airween replied:
Thanks, now it works as well both with mod_security2 (Apache) and libmodsecurity3 + nginx.

azurit commented:
Simple way of testing anti-virus:
`python -c 'import requests; print(requests.post("http://example.com/", files={"virus": "X5O!P%@AP[4\PZX54(P^)7CC)7}$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!$H+H*"}))'`

azurit commented:
@airween Thanks A LOT for a detailed review!

azurit commented:
Just a note: I wil rewrite this into plugin after plugin infrastructure is merged.

dune73 commented:
Plugin infrastructure is awaiting review by @lifeforms. Likely next week. It should be easy to adopt for you.

emphazer commented:
this PR is very interesting! I will review it too with modsec 2. pls don't merge it before I checked all kind of POST and PUT methods and different types of body tmp files in Apache top folder. 
I'm sure I will be done with the tests this week!

azurit commented:
> hope that's okay?!?

Sure! This won't me merged either way as it's going to be rewritten into plugin. But feel free to review it, functionality and Lua script will, probaby, stay without any changes.

emphazer commented:
@azurit 
i'm done with testing. 

I tested with Modsec v2.93, Apache 2.4, Redhat EL7, LUA 5.1 and lua-socket 3.0.

I found several Issues.


Problem 1:
The LUA script doesn't support 2 of 3 common upload methods.
1. POST with _Content-Type: application/x-www-form-urlencoded_
     curl -X POST -d @eicar.txt http://localhost/
2. PUT 
    curl -T eicar.txt http://localhost/

i think the scan engine skips the files because the tmp files are empty if you use this upload methods.
the main idea how you handle the file scanning here is the same like in the open source lua script https://www.angeloxx.co.uk/technews-en/clamav-lua-script-for-modsecurity.html

i know that problem already from the past.
we have two different tmp files depending on the request method.
the common tmp file you scan has a name like that 20210406-191809-YGyX0YBsplWVOR-Avd65WAAAANc-file-yQpu28.
but if you use an urlencoded POST then the filename looks like file-YGyX0YBsplWVOR-Avd65WAAAANc.tmp.
i'm not sure but i guess that this file is generated by apache and not modsecurity.
i used SecUploadFileMode 0660 and the other file has 644 on the filesystem.

2 years ago i modified the opensource lua script to handle all 3 methods with a little dirty hack. i scan all files with the uniq modsecurity id in the filename. *YGyX0YBsplWVOR-Avd65WAAAANc*
it looks like that _/tmp/*" .. m.getvar("unique_id") .. "*"_

i know it is a dirty hack but it works.


Problem 2: 
the legacy support for LUA 5.1 socket doesn't work for me. i spend 2 hours yesterday and tried really everything.
tcp works fine.

the error message is: 
Message: Rule processing failed (id=, msg=).
Apache-Error: [file "apache2_util.c"] [line 271] [level 3] [client 127.0.0.1] ModSecurity: Rule processing failed (id=, msg=). [hostname "localhost"] [uri "/"] [unique_id "YG1QtE@RLL-cN21TfIZ8TgAAAkg"]

i couldn't try it with a newer LUA version because we use just backported distribution packages maintained by Redhat.

if you want i can send you the lua script if you give me your email. 
but i have to warn you... i'm not a lua developer or programer at all ;-), but maybe you can use it as inspiration to fix this problem. i think it is really important that we support all methods.

i have no clue if with nginx or modsec v3 this problem exists too. i hope that anybody else can test it.

dune73 commented:
Yes, same problem here, but I did not return to it unfortunately despite guidance by @azurit. Too many other tasks ...

azurit commented:
> i think the scan engine skips the files because the tmp files are empty if you use this upload methods.
> the main idea how you handle the file scanning here is the same like in the open source lua script https://www.angeloxx.co.uk/technews-en/clamav-lua-script-for-modsecurity.html

No, i'm using completely differeny approach. This script is scannig by running external utility called `clamscan`.

emphazer commented:
@azurit i know. your script is totally different and way better.

i'm just talking about the mothod to open the file.
this line here:
local file_handle = io.open(filename, "r")

azurit commented:
Ok, i will look into it BUT i'm getting filename and path from modsecurity (it is sending it as first argument to Lua script).

airween commented:
Hi @emphazer,

> Problem 1:
> The LUA script doesn't support 2 of 3 common upload methods.
> 
>     1. POST with Content-Type: application/x-www-form-urlencoded
>        curl -X POST -d @eicar.txt http://localhost/
>     2. PUT
>        curl -T eicar.txt http://localhost/
> 
> i think the scan engine skips the files because the tmp files are empty if you use this upload methods.

yes, IMHO FILES* collections will filled only if the request type is `multipart/form-data`.

Just my 2 cents: as I see, [PUT](https://github.com/coreruleset/coreruleset/blob/v3.4/dev/rules/REQUEST-901-INITIALIZATION.conf#L178-L185) [isn't allowed](https://github.com/coreruleset/coreruleset/blob/v3.4/dev/rules/REQUEST-911-METHOD-ENFORCEMENT.conf#L27-L28) by default. But it's no problem of course - may be we should notice users in comment.

> Problem 2:
> the legacy support for LUA 5.1 socket doesn't work for me. i spend 2 hours yesterday and tried really everything.
> tcp works fine.

you wrote that you're using RHEL7 - isn't there any SELinux settings, which blocks the access to the socket?

> if you want i can send you the lua script if you give me your email.

can you send me that script too? :)

Thanks.

emphazer commented:
@azurit 
here is the hack i did

     local filename = "/tmp/file-" .. m.getvar("unique_id") .. ".tmp"
     local filehandle=assert(io.open(filename,"w+b"))
     filehandle:write(m.getvar("REQUEST_BODY"))
     filehandle:close()

emphazer commented:
> 
> 
> Ok, i will look into it BUT i'm getting filename and path from modsecurity (it is sending it as first argument to Lua script).

yeah, i used SecRuleScript instead. but it should also work with  FILES_TMPNAMES and @inspectFile

azurit commented:
> * POST with _Content-Type: application/x-www-form-urlencoded_
>   curl -X POST -d @eicar.txt http://localhost/


This isn't a file upload. I'm not sure if we should support virus scanning of `REQUEST_BODY` as it will scan all POST requests so it will be really expensive. Maybe as an optional feature but it will be crazy and slow to send everything into anti-virus software.


> * PUT
>   curl -T eicar.txt http://localhost/

This isn't a file upload, too. Even more, i don't see any data beeing send using this - just use `-v` switch to curl.

azurit commented:
Ok, in case of `PUT`, i would allow to optionally allow data scanning.

emphazer commented:
@azurit 
> 
> > * POST with _Content-Type: application/x-www-form-urlencoded_
> >   curl -X POST -d @eicar.txt http://localhost/
> 
> This isn't a file upload. I'm not sure if we should support virus scanning of `REQUEST_BODY` as it will scan all POST requests so it will be really expensive. Maybe as an optional feature but it will be crazy and slow to send everything into anti-virus software.
> 
i work for the gouvernement and that was exactly what they wanted. we did benchmark with PUT and POST here. everything was fine with clamdscan as long you use it as a demon and not as commandline scanner.

if you could do it for the urlencoded POST as additional security variable for high risk systems, which is per default off, it would be absolutly fantastic!!! :-)

azurit commented:
Hey @emphazer, I was successfull in fullfilling your wishes :D Try it now but don't forget to check new setting in `crs-setup.conf`.

emphazer commented:
works in tcp mode excellent!
you are a genius ;-)

azurit commented:
Cool!

What about socket file? Can you post your anti-virus configuration from `crs-setup.conf`?

emphazer commented:
> 
> 
> Are you sure your modsecurity is compiled with Lua 5.3? Is it modsec 2 or 3? The whole modsecurity 2.x documentation is mentioning only Lua version 5.1. Newer Lua versions are available only with modsec 3 (as far as i understand it).

that are really really bad news 

@dune73 does your mod_security show compiled with LUA 5.3 if you start the apache?

emphazer commented:
Message: Rule processing failed (id=925110, msg=Virus detected in uploaded file: %{tx.virus_name}).

same message if i stop the clamd service

SecAction \
 "id:900800,\
  phase:1,\
  nolog,\
  pass,\
  t:none,\
  setvar:'tx.crs_anti-virus_protection_enable=1',\
  setvar:'tx.crs_anti-virus_protection_scan_request_body=1',\
  setvar:'tx.crs_anti-virus_clamav_connect_type=socket',\
  setvar:'tx.crs_anti-virus_clamav_socket_file=/var/run/clamd.scan/clamd.sock',\
  setvar:'tx.crs_anti-virus_clamav_address=127.0.0.1',\
  setvar:'tx.crs_anti-virus_clamav_port=3310',\
  setvar:'tx.crs_anti-virus_clamav_chunk_size=4096',\
  setvar:'tx.crs_anti-virus_clamav_network_timeout_seconds=2',\
  setvar:'tx.crs_anti-virus_clamav_max_file_size_bytes=1048576'"

azurit commented:
Please do these and post the results of all (set the correct path to your ClamAV config file):
`ls -la /var/run/clamd.scan/clamd.sock`
`grep LocalSocket /etc/clamav/clamd.conf`
`clamdscan --stream --config-file=/etc/clamav/clamd.conf eicar.txt`


Thank you!

emphazer commented:
```
ll /var/run/clamd.scan/ -a
total 0
drwxr-xr-x.  2 clamscan virusgroup   60 Apr  9 10:12 .
drwxr-xr-x. 45 root     root       1580 Apr  9 09:56 ..
srw-rw-rw-.  1 clamscan virusgroup    0 Apr  9 10:12 clamd.sock

LocalSocket /var/run/clamd.scan/clamd.sock
LocalSocketGroup virusgroup
LocalSocketMode 666

/tmp/eicar.txt: Eicar-Signature FOUND

----------- SCAN SUMMARY -----------
Infected files: 1
Time: 0.006 sec (0 m 0 s)
Start Date: 2021:04:09 10:16:49
End Date:   2021:04:09 10:16:49
```

azurit commented:
Looks ok. If you have some time, we will have to debug Lua script and find out where exactly it is failing.

You can use this code to log messages into apache error log:
`m.log(2, 'message')`

Try logging, for example, different numbers on various places in the anti-virus Lua script, like this:
```
m.log(2, 'DEBUG: 1')
...
m.log(2, 'DEBUG: 2')
...
m.log(2, 'DEBUG: 3')
```

Try adding debug messages at least at these lines ( https://github.com/coreruleset/coreruleset/blob/2c7f3b4104f76367a27d9f1b41e6758b4f240d2b/rules/anti-virus.lua ):
 - before and after line 24
 - before and after line 26
 - before and after line 45
 - before and after line 46
 - before and after line 49
 - before and after line 55

Don't forget to reload apache after doing any changes in the Lua script.

emphazer commented:
> m.log(2, 'message')

okay, where do i have to place this line?

azurit commented:
> Try adding debug messages at least at these lines ( https://github.com/coreruleset/coreruleset/blob/2c7f3b4104f76367a27d9f1b41e6758b4f240d2b/rules/anti-virus.lua ):
> 
>     * before and after line 24
> 
>     * before and after line 26
> 
>     * before and after line 45
> 
>     * before and after line 46
> 
>     * before and after line 49
> 
>     * before and after line 55

Here ---^ but log different message every time so we will see where exactly script stopped.

azurit commented:
```
m.log(2, 'DEBUG: 1')
local ok, socket = pcall(require, module_name)
m.log(2, 'DEBUG: 2')
...
m.log(2, 'DEBUG: 3')
m.log(2, "ClamAV: lua-socket library not installed, please install it or disable 'tx.crs_anti-virus_protection_enable' in crs-setup.conf.")
m.log(2, 'DEBUG: 4')
...
m.log(2, 'DEBUG: 5')
sck = socket.tcp()
m.log(2, 'DEBUG: 6')
...
m.log(2, 'DEBUG: 7')
sck:settimeout(tonumber(m.getvar("tx.crs_anti-virus_clamav_network_timeout_seconds", "none")))
m.log(2, 'DEBUG: 8')
...
m.log(2, 'DEBUG: 9')
status, error = sck:connect(m.getvar("tx.crs_anti-virus_clamav_socket_file"))
m.log(2, 'DEBUG: 10')
...
m.log(2, 'DEBUG: 11')
m.log(2, string.format("ClamAV: Error connecting to antivirus: %s", error))
m.log(2, 'DEBUG: 12')
```

emphazer commented:
okay, got it.
i got DEBUG 1, 2 and 5.

it pass line 24 and stops at line 45

azurit commented:
What is your version of Lua and lua-socket library?

emphazer commented:
I tested with Modsec v2.93, Apache 2.4, Redhat EL7, LUA 5.1.4 and lua-socket 3.0.

azurit commented:
Try running this lua script outside of the apache:
```
local ok, socket = pcall(require, "socket.unix")
sck = socket.tcp()
status, error = sck:connect("/var/run/clamd.scan/clamd.sock")
print(status)
print(error)
sck:close()
```

Save it, for example, into file `test.lua` and run it with `lua test.lua`.

azurit commented:
Also check if you are running correct version of the Lua (the same as in apache): `lua -v`.

emphazer commented:
Lua 5.1.4  Copyright (C) 1994-2008 Lua.org, PUC-Rio

azurit commented:
What about that script in https://github.com/coreruleset/coreruleset/pull/1986#issuecomment-816557240? What happened when you run it on command line outside apache?

emphazer commented:
```
lua: test.lua:2: attempt to index local 'socket' (a function value)
stack traceback:
        test.lua:2: in main chunk
        [C]: ?
```

azurit commented:
Try this one:
```
local ok, socket2 = pcall(require, "socket.unix")
sck = socket2.tcp()
status, error = sck:connect("/var/run/clamd.scan/clamd.sock")
print(status)
print(error)
sck:close()

```

emphazer commented:
```
lua: test2.lua:2: attempt to index local 'socket2' (a function value)
stack traceback:
        test2.lua:2: in main chunk
        [C]: ?
```

azurit commented:
That's really weird, everything is fully working with same versions of the software on Debian 10.

What about this:
```
local ok, socket = pcall(require, "socket.unix")
print(ok)
print(type(socket))
print(socket)
sck = socket.tcp()
status, error = sck:connect("/var/run/clamd.scan/clamd.sock")
print(status)
print(error)
sck:close()
```

emphazer commented:
i don't understand it too. the lua-socket module is from epel repository.
and lua from official rhel7 repo...

```
true
function
function: 0x60c080
lua: test3.lua:5: attempt to index local 'socket' (a function value)
stack traceback:
        test3.lua:5: in main chunk
        [C]: ?
```

azurit commented:
Can you point me to the repository page of the lua-socket you are using? Looks like you have something different than i have. You need this library:
http://w3.impa.br/~diego/software/luasocket/

emphazer commented:
```
Installed Packages
Name        : lua-socket
Arch        : x86_64
Version     : 3.0
Release     : 0.17.rc1.el7
Size        : 671 k
Repo        : installed
From repo   : rhel-server-7-epel
Summary     : Network support for the Lua language
URL         : http://www.tecgraf.puc-rio.br/~diego/professional/luasocket/
License     : MIT
Description : LuaSocket is a Lua extension library that is composed by two parts: a C core
            : that provides support for the TCP and UDP transport layers, and a set of Lua
            : modules that add support for functionality commonly needed by applications
            : that deal with the Internet.
            :
            : Among the support modules, the most commonly used implement the SMTP, HTTP
            : and FTP. In addition there are modules for MIME, URL handling and LTN12.
```



here the link to the source package. last update was 2018
https://ftp.fau.de/epel/7/SRPMS/Packages/l/lua-socket-3.0-0.17.rc1.el7.src.rpm

emphazer commented:
@azurit you use v3.0 too right?

emphazer commented:
well, i could recompile the src package without the 3 included patches and see if it works...
or else im out of ideas.

azurit commented:
No, wait with that please, i will try to look at it again probably tomorrow (i'm out of energy for today). Thank you!

emphazer commented:
no problemo ;-)

airween commented:
I don't know why, but looks like the RH and for eg. Debian Luasocket are different (I've tested both of them).

Here is a snippet which worked on both system:

```lua
socket = require("socket")
socket.unix = require("socket.unix")
c = socket.unix()
local ok = c:connect("/var/run/clamav/clamd.ctl")
print(ok)
```

Also it's interesting that on Debian the result will `1`, on RH will `1.0`.

azurit commented:
@airween Thanks!

@emphazer Can you try current version? Pls test both socket and tcp.

emphazer commented:
yeeeeeehaaaa, work perfect! :-)
good job!

azurit commented:
Does ANYONE see any problems or have feature requests?

airween commented:
> Does ANYONE see any problems or have feature requests?

I think it's really nice work, congratulation!

I'd check the whole cumulated patch again both on Apache and Nginx, but may be I just can do that tomorrow.

dune73 commented:
This is coming along nicely. Thank you all for your work and looking fwd to @airween's review.

airween commented:
@azurit, @emphazer - can you help me how can I emulate the `POST` request with `x-www-form-urlencoded` CT?

After the first tests, this implementation works with the "regular" file upload:

```
curl -v http://localhost/index.php -F "userfile=@eicar.com"
```

```
2021/04/17 13:42:10 [info] 17102#17102: *11 ModSecurity: Warning. Matched "Operator `StrEq' with parameter `' against variable `TX:VIRUS_NAME' (Value: `Win.Test.EICAR_HDB-1' ) [file "/usr/share/modsecurity-crs/rules/REQUEST-925-ANTIVIRUS.conf"] [line "30"] [id "925110"] [rev ""] [msg "Virus detected in uploaded file: Win.Test.EICAR_HDB-1"] [data "Virus detected in uploaded file: Win.Test.EICAR_HDB-1"] [severity "2"] [ver "OWASP_CRS/3.3.0"] [maturity "0"] [accuracy "0"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/262/441/442"] [hostname "::1"] [uri "/index.php"] [unique_id "1618659730"] [ref "v286,63"], client: ::1, server: _, request: "POST /index.php HTTP/1.1", host: "localhost"

2021/04/17 13:42:10 [info] 17102#17102: *11 ModSecurity: Warning. Matched "Operator `StrEq' with parameter `' against variable `TX:VIRUS_NAME' (Value: `Win.Test.EICAR_HDB-1' ) [file "/usr/share/modsecurity-crs/rules/REQUEST-925-ANTIVIRUS.conf"] [line "46"] [id "925120"] [rev ""] [msg "Virus detected in request body: Win.Test.EICAR_HDB-1"] [data "Virus detected in request body: Win.Test.EICAR_HDB-1"] [severity "2"] [ver "OWASP_CRS/3.3.0"] [maturity "0"] [accuracy "0"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/262/441/442"] [hostname "::1"] [uri "/index.php"] [unique_id "1618659730"] [ref "v182,389"], client: ::1, server: _, request: "POST /index.php HTTP/1.1", host: "localhost"
```

and PUT method:

```
curl -v -T "eicar.com" "http://localhost/index.php"
```

```
2021/04/17 13:39:00 [info] 17102#17102: *9 ModSecurity: Warning. Matched "Operator `StrEq' with parameter `' against variable `TX:VIRUS_NAME' (Value: `Win.Test.EICAR_HDB-1' ) [file "/usr/share/modsecurity-crs/rules/REQUEST-925-ANTIVIRUS.conf"] [line "46"] [id "925120"] [rev ""] [msg "Virus detected in request body: Win.Test.EICAR_HDB-1"] [data "Virus detected in request body: Win.Test.EICAR_HDB-1"] [severity "2"] [ver "OWASP_CRS/3.3.0"] [maturity "0"] [accuracy "0"] [tag "paranoia-level/1"] [tag "OWASP_CRS"] [tag "capec/1000/262/441/442"] [hostname "::1"] [uri "/index.php"] [unique_id "1618659540"] [ref "v117,184"], client: ::1, server: _, request: "PUT /index.php HTTP/1.1", host: "localhost"
```

But the rule 925120 didn't triggered with the request was like this:
```
curl -X POST -d @eicar.com http://localhost/index.php
```

And my 2 cents: in case of first case (regular file upload) there are two blocking rules. It's no problem, just saying.

azurit commented:
> But the rule 925120 didn't triggered with the request was like this:

It works for me. Is that modsec3? Have you set `tx.crs_anti-virus_protection_scan_request_body` to `1` in `crs-setup.conf`?

> And my 2 cents: in case of first case (regular file upload) there are two blocking rules. It's no problem, just saying.

modsec3?

airween commented:
> It works for me.

Can you share with me the `curl` request for this test?

> Is that modsec3? Have you set `tx.crs_anti-virus_protection_scan_request_body` to `1` in `crs-setup.conf`?

yes, and yes :)

But I also tested with mod_security2 with **same result**.

> > And my 2 cents: in case of first case (regular file upload) there are two blocking rules. It's no problem, just saying.
> 
> modsec3?

yes, but in case of mod_security2, the `PUT` request doesn't work at there too. Can you share that request too?

azurit commented:
I used the same command as you:
`curl -X POST -d @eicar.com http://localhost/index.php`

azurit commented:
Apache or nginx?

airween commented:
I'm using Apache2 with mod_security2, and Nginx with libmodsecurity3.

Here is the debug log from Apache+mod_security2 and the request what you shared:

```
Recipe: Invoking rule 7f928abeec38; [file "/usr/share/modsecurity-crs/rules/REQUEST-925-ANTIVIRUS.conf"] [line "58"] [id "925120"].
Rule 7f928abeec38: SecRule "TX:CRS_ANTI-VIRUS_PROTECTION_SCAN_REQUEST_BODY" "@eq 1" "phase:2,log,auditlog,id:925120,block,t:none,msg:'Virus detected in request body: %{tx.virus_name}',logdata:'Virus detected in request body: %{tx.virus_name}',tag:paranoia-level/1,tag:OWASP_CRS,tag:capec/1000/262/441/442,ver:OWASP_CRS/3.3.0,severity:CRITICAL,chain"
Transformation completed in 1 usec.
Executing operator "eq" with param "1" against TX:crs_anti-virus_protection_scan_request_body.
Target value: "1"
Operator completed in 1 usec.
Rule returned 1.
Match -> mode NEXT_RULE.
Recipe: Invoking rule 7f928abe4a88; [file "/usr/share/modsecurity-crs/rules/REQUEST-925-ANTIVIRUS.conf"] [line "59"].
Rule 7f928abe4a88: SecRule "REQUEST_BODY" "@inspectFile anti-virus.lua" "chain"
Transformation completed in 1 usec.
Executing operator "inspectFile" with param "anti-virus.lua" against REQUEST_BODY.
Target value: "PK\x03\x04"
Lua: Executing script: /usr/share/modsecurity-crs/rules/anti-virus.lua
Lua: Script completed in 1600 usec, returning: (null).
Operator completed in 1622 usec.
Rule returned 0.
```

Of course, I upgraded the `anti-virus.lua` tool by your newest version.

But looks like `curl` sends only a part of the file:
```
--37d8e57b-C--
PK^C^D
--37d8e57b-F--
```
Thoughts?

airween commented:
I reviewed the whole PR again both on Apache2 + mod_security2 and Nginx and libmodsecurity3.

There were two files: `eicar.com`, which is a **zip** file, and `eicar.txt` the plain virus itself.

I tested the engines with these requests:
```
curl http://localhost/index.php -F "userfile=@eicar.com"
curl http://localhost/index.php -F "userfile=@eicar.txt" 
curl http://localhost/index.php -X POST -d "@eicar.txt"
curl http://localhost/index.php -T "eicar.txt"
```

the requests:

1. upload the zip file
2. upload the plain virus
3. send a POST request with CT `x-www-form-urlencoded`
4. send a PUT request

All request were catched by both engine, the one and only side-effect is when I uploaded the zipped file to Nginx, it logged the both rule ([925110](https://github.com/azurit/coreruleset/blob/v3.4/antivirus/rules/REQUEST-925-ANTIVIRUS.conf#L30-L44) and [925120](https://github.com/azurit/coreruleset/blob/v3.4/antivirus/rules/REQUEST-925-ANTIVIRUS.conf#L46-L61)) - the reason is known (`REQUEST_BODY` is always exists in v3).

I asked @azurit that fix the prerequisites in case of v3 and nginx in comment - every other part of this PR is perfect :)

azurit commented:
@airween Thank you very much!

Can you be so kind and test again - this time as a plugin, see:
https://github.com/coreruleset/antivirus-plugin

Don't forget to apply changes in includes needs for plugins to work:
https://github.com/coreruleset/coreruleset/blob/v3.4/dev/INSTALL

Almost nothing changed so do only quick test mainly on Nginx and modsec3 (i already tested modsec2 + apache).

airween commented:
> @airween Thank you very much!
> 
> Can you be so kind and test again - this time as a plugin, see:
> https://github.com/coreruleset/antivirus-plugin
> 
> Don't forget to apply changes in includes needs for plugins to work:
> https://github.com/coreruleset/coreruleset/blob/v3.4/dev/INSTALL
> 
> Almost nothing changed so do only quick test mainly on Nginx and modsec3 (i already tested modsec2 + apache).

sure, I'll check it again soon.

azurit commented:
Closing in favour of official Antivirus plugin, see:
https://github.com/coreruleset/antivirus-plugin

Thanks to everyone!
